universe/ValueRefs.cpp

```diff
@@ -1322,6 +1331,83 @@ std::string Statistic<std::string, std::string>::Eval(const ScriptingContext& co
     return max->first;
 }
 
+///////////////////////////////////////////////////////////
+// TotalFighterShots (of a carrier during one battle)    //
+///////////////////////////////////////////////////////////
+TotalFighterShots::TotalFighterShots(std::unique_ptr<Condition::Condition>&& sampling_condition) :
```

Comment:
Seems unliklely that you need a whole extra ValueRef class for this, and if you do, it's probably a more general concept than what you've called it here.

Reply:
I'm also unclear why you need the sampling condition to be passed in and can't get it from parts in the design...

And the design of this ValueRef is in general odd... it seems like it should be a property of the ship and accessible with `Source.TotalFighterShots` or similar as a bound property value ref, and not a separate type of reference. It's not clear from this design what object's properties are actually being accessed...

Reply:
I am not sure it is clear what this about (see next comment). 

In principle, if fighter-launches were universe objects parameterized with the combat bout one could sum those up with a statistic. But that would be much more horrible.

Reply:
I meant use the `Statistic` framework to specify a ValueRef with a condition parameter, and add a TotalFighterShots statistic type and move all the related logic into the appropriate case when evaluating. But I guess the logic being applied doesn't really use the condition like a statitic would (to filter objects to evalute the valueref on...)

Reply:
> I'm also unclear why you need the sampling condition to be passed in and can't get it from parts in the design...

It was intended to use a simplified sampling condition as opposed to the combatTargets (e.g. visibility checks can/should be omitted for damage estimation).

> it seems like it should be a property of the ship and accessible with `Source.TotalFighterShots` or similar as a bound property value ref, and not a separate type of reference.

If the sampling condition is separate, having it as a bound property value ref is not really feasible (?).

> It's not clear from this design what object's properties are actually being accessed...

Yes, it is only in the documentation. 

Maybe rename to SourcesTotalFighterShots? or add a carrier id parameter?

or should I unpower it and always use the slower combatTargets condition? 

it could also be decoupled from a ship; you would need to pass in launch capacity, hangar capacity and sampling condition.

Reply:
Looking at it again, it could possibly be implemented as a Statistic. The condition would be your combatTargets condition, and the single value could be the ID of the ship to inspect. You'd then need a new statistic type for this specific use, which would be handled differently in `Statistic<T, V>::Eval` at the start, by not doing a search for condition matches or a call to `ReduceData<T, V>`.

Not sure if that's worth the complication / limits, though...

Reply:
> Looking at it again, it could possibly be implemented as a Statistic. 

Would you imagine usage like this?
```
totalFighterDamageEstimation = Statistic TotalFighterShots value = Source.ID condition = And [ [[COMBAT_TARGETS_VISIBLE_ENEMY]] Fighter ]
```
I do not see what is the benefit of doing that. It would be more flexible, because it could technically be passed more than a single object, but then actually it is probably a sum/needs a statistic type.

Also the misuse of the value parameter would be quite confusing.

The following would be probably closer to usual Statistic usage, but still calculating a Sum over multiple bouts instead of over multiple filtered objects 
```
totalFighterDamageEstimation = CombatStatistic Sum value = Source.LaunchedFightersCount condition = And [ [[COMBAT_TARGETS_VISIBLE_ENEMY]] Fighter ]
```
Note this assumes there is at maximum one candidate, else one would need something like  `...condition = NumberOf number = 1 condition = And ...`